node/node_cross_shard.go

```diff
@@ -120,124 +119,75 @@ func (node *Node) VerifyBlockCrossLinks(block *types.Block) error {
 		)
 	}
 
-	firstCrossLinkBlock := core.EpochFirstBlock(node.Blockchain().Config().CrossLinkEpoch)
-
-	for i, crossLink := range *crossLinks {
-		lastLink := &types.CrossLink{}
-		if i == 0 {
-			if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 {
-				lastLink, err = node.Blockchain().ReadShardLastCrossLink(crossLink.ShardID())
-				if err != nil {
-					return ctxerror.New("[CrossLinkVerification] no last cross link found 1",
-						"blockHash", block.Hash(),
-						"crossLink", lastLink,
-					).WithCause(err)
-				}
-			}
-		} else {
-			if (*crossLinks)[i-1].Header().ShardID() != crossLink.Header().ShardID() {
-				if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 {
-					lastLink, err = node.Blockchain().ReadShardLastCrossLink(crossLink.ShardID())
-					if err != nil {
-						return ctxerror.New("[CrossLinkVerification] no last cross link found 2",
-							"blockHash", block.Hash(),
-							"crossLink", lastLink,
-						).WithCause(err)
-					}
-				}
-			} else {
-				lastLink = &(*crossLinks)[i-1]
-			}
-		}
-
-		if crossLink.BlockNum().Cmp(firstCrossLinkBlock) > 0 { // TODO: verify genesis block
-			err = node.VerifyCrosslinkHeader(lastLink.Header(), crossLink.Header())
-			if err != nil {
-				return ctxerror.New("cannot ValidateNewBlock",
+	for _, crossLink := range *crossLinks {
+		cl, err := node.Blockchain().ReadCrossLink(crossLink.ShardID(), crossLink.BlockNum())
+		if err == nil && cl != nil {
+			if !bytes.Equal(cl.Serialize(), crossLink.Serialize()) {
+				return ctxerror.New("[CrossLinkVerification] Double signed crossLink",
```

Comment:
When there is a double sign, add a TODO here. Don't return error, then double-sign will fail the system.

Reply:
It's still returning error? If there is really a cross-link with double sign, it's a proof for slashing. not an invalid block.